src/linker/Linker.Steps/RootAssemblyInputStep.cs

```diff
@@ -85,14 +97,24 @@ AssemblyDefinition LoadAssemblyFile ()
 			return assembly;
 		}
 
-		void MarkAndPreserveVisible (TypeDefinition type, TypePreserveMembers preserve)
+		void MarkAndPreserve (AssemblyDefinition assembly, TypePreserveMembers preserve)
 		{
-			switch (preserve) {
-			case TypePreserveMembers.AllVisible when !IsTypeVisible (type):
-				return;
-			case TypePreserveMembers.AllVisibleOrInternal when !IsTypeVisibleOrInternal (type):
-				return;
-			}
+			var module = assembly.MainModule;
+			if (module.HasExportedTypes)
+				foreach (var type in module.ExportedTypes)
+					MarkAndPreserve (assembly, type, preserve);
+
+			foreach (var type in module.Types)
+				MarkAndPreserve (type, preserve);
+		}
+
+		void MarkAndPreserve (TypeDefinition type, TypePreserveMembers preserve)
+		{
+			if ((preserve & TypePreserveMembers.Visible) != 0 && !IsTypeVisible (type))
+				preserve &= ~TypePreserveMembers.Visible;
+
+			if ((preserve & TypePreserveMembers.Internal) != 0 && !IsTypeVisibleOrInternal (type))
+				preserve &= ~TypePreserveMembers.Internal;
```

Comment:
`IsTypeVisibleOrInternal` doesn't feel right here. The current check is basically anything but private, but to match the new semantics of the enum, it should be "internal".

Reply:
Resolved